### ISSUE TITLE
Can't able to reproduce results

### ISSUE THREAD

**Issue by Karn3003** [KozakHou/Python_Tensorflow_Fourier_Neural_Operator#1]
Hi, I am unable to reproduce the results reported in the repository. Also Can you provide the logic behind the Fourier Layer you used as I am confused with implementation details.

**Comment by KozakHou** [comment]
Hi, first of all, thank you for your response. This time, I used a different remote terminal (GPU: RTX 6000) to run the FNO_MNIST_A100.ipynb file, and it was able to reproduce almost the same results.

I suspect that the decrease in batch size may have contributed to this. While reducing the batch size can reduce memory consumption, it typically affects the performance of the model. Based on my personal experiments, I found that the batch size should be at least 128 or higher to achieve similar results.

I believe the underlying principle of FNO is related to High Boost Filtering in image processing. First, the image needs to be projected onto the Fourier eigenspace to obtain signals ranging from high to low frequencies. The idea is to remove the high-frequency signals, which extracts the edges of the image, similar to a smoothing filter. Then, the original image is added back to strengthen the low-frequency signals without losing the high-frequency signals. This concept is similar to ResNet, where the model weights are improved to enhance accuracy. I think if FNO is used for image processing or classification, this is its underlying concept.

If you have any other questions, please feel free to ask. Thank you.

**Comment by Karn3003** [comment]
Okay, I will try different batch sizes. Still, in Original Paper after removing the high frequency they do linear transformation (parameters are learnable in the spatial domain) and then do the inverse Fourier transform and add original input via linear transform and add like skip connection. In your code for the Fourier block, I see you are eliminating the higher frequency and then multiple the original signal and adding back the original input which is a quite different approach mentioned in the paper as they were doing spatial convolution. Can you let me know your thoughts on doing that wrt to the following code you wrote I am a little bit confused reading what you want to do here I understood everything else apart from the following.

    out_ft_first = tf.Variable(np.array(out_ft.shape))
    out_ft_first = tf.math.multiply(X[:, :, :4, :4], tf.cast(Res_X[:, :, :4, :4], tf.complex64))

    out_ft_end  = tf.Variable(np.array(out_ft.shape))
    out_ft_end = tf.math.multiply(X[:, :, -4:, :4], tf.cast(Res_X[:, :, -4:, :4], tf.complex64))


    o1 = tf.concat([out_ft_first,out_ft[:,:,:4,4:]],axis=3)
    o2 = tf.concat([out_ft[:,:,-4:,4:],out_ft_end],axis=3)
    
    out_ft = tf.concat([o1,o2],axis=2)
    
Thanks for your reply.

**Comment by KozakHou** [comment]
I remember when implementing this piece of code, I encountered an issue of not being able to specify how much high-frequency signal should be eliminated. Therefore, I first set up an out_ft and made it conform to the input matrix size.

I also encountered a problem where I couldn't perform tf.math.multiply() without first setting up out_ft_first=tf.Variable() and out_ft_end = tf.Variable(). This was a confusion I had during implementation.

In summary, the first step is to transform Res_X into the complex domain. Then, multiply it separately with X[:, :, :4, :4] and X[:, :, -4, :4]. This part was inspired by Zongyi Li's approach in simulating Darcy Flow with FNO. As I understand it, this direction allows the resulting out_ft_first and out_ft_end to become a Low-Pass Filter.

For a visual explanation, please refer to the images in this URL (https://www.alluxa.com/optical-filter-specs/spectral-features/).

I think it can be understood that o1 and o2 represent the areas under a Gaussian function, skewed towards the left and right halves respectively. out_ft is then used to set the values outside a distance of 4 units from the center to 0 (out_ft = tf.zeros()), and finally, the variable out_ft is replaced with o1 and o2 concatenated together. This results in a matrix with only low-frequency components, which is then inverted.

I believe there must be simpler ways to implement a low-pass filter, and the concatenation approach I used may be confusing. Since the output X must match the size of the original Res_X matrix, it involved some non-intuitive operations, including concatenation with a matrix of all zeros, to ensure smooth execution of matrix operations.

I also hope to develop a more efficient Fourier Block. Let's work together towards that goal!

**Comment by Karn3003** [comment]
Thanks for the reply. Yeah sure, let's work towards more efficient FNO blocks. I think it might be able to help you as I already went through the whole code base written. 

https://github.com/neuraloperator/neuraloperator

**Comment by Karn3003** [comment]
Hi, Just one Last question I have. I went through the original Implementation code base and had doubts about following the code for 2d Spectral Convolution as what they are trying to do. It would be great if you get it and explain it to me.

class FactorizedSpectralConv2d(FactorizedSpectralConv):
    def forward(self, x, indices=0):
        batchsize, channels, height, width = x.shape

        x = torch.fft.rfft2(x.float(), norm=self.fft_norm)

        # The output will be of size (batch_size, self.out_channels, x.size(-2), x.size(-1)//2 + 1)
        out_fft = torch.zeros([batchsize, self.out_channels, height, width//2 + 1], dtype=x.dtype, device=x.device)

        # upper block (truncate high freq)
        out_fft[:, :, :self.half_n_modes[0], :self.half_n_modes[1]] = self._contract(x[:, :, :self.half_n_modes[0], :self.half_n_modes[1]], 
                                                                              self._get_weight(2*indices), separable=self.separable)
        # Lower block
        out_fft[:, :, -self.half_n_modes[0]:, :self.half_n_modes[1]] = self._contract(x[:, :, -self.half_n_modes[0]:, :self.half_n_modes[1]],
                                                                              self._get_weight(2*indices + 1), separable=self.separable)
        
        if self.output_scaling_factor is not None:
            width = int(round(width*self.output_scaling_factor[0]))
            height = int(round(height*self.output_scaling_factor[1]))

        x = torch.fft.irfft2(out_fft, s=(height, width), dim=(-2, -1), norm=self.fft_norm)

        if self.bias is not None:
            x = x + self.bias[indices, ...]

        return x
        
  Thanks.
